### ISSUE TITLE
minetestserver crashes following random CPU spike

### ISSUE THREAD

**Issue by acoolstraw** [minetest/minetest#12487]
##### Minetest version

```
Minetest 5.5.1 (Linux)
Using LuaJIT 2.1.0-beta3
BUILD_TYPE=Release
RUN_IN_PLACE=0
USE_CURL=1
STATIC_SHAREDIR="/usr/share/minetest"
```

##### OS / Hardware

Operating system: Arch Linux x86_64
CPU: Intel (Haswell, no TSX) (2) @ 2.599GHz

(it's a VPS running Arch Linux)

##### Summary

At very random times (sometimes when exploring ungenerated chunks, sometimes just completely out of nowhere), minetestserver crashes with the following logs from `systemctl status minetest@minetest`:
```
Jun 28 21:12:48 one minetestserver[362992]: 2022-06-28 21:12:48: WARNING[Server]: Deprecated usage of get_node_group, use get_item_group instead (at ...etest/games/mineclone5/mods/ITEMS/mcl_core/fun>
Jun 28 21:12:48 one minetestserver[362992]: 2022-06-28 21:12:48: ACTION[Server]: cy moves stuff to chest at (-238,7,-145)
Jun 28 21:12:49 one minetestserver[362992]: 2022-06-28 21:12:49: WARNING[Server]: Deprecated usage of get_node_group, use get_item_group instead (at ...etest/games/mineclone5/mods/ITEMS/mcl_core/fun>
Jun 28 21:12:50 one minetestserver[362992]: 2022-06-28 21:12:50: WARNING[Server]: Deprecated usage of get_node_group, use get_item_group instead (at ...etest/games/mineclone5/mods/ITEMS/mcl_core/fun>
Jun 28 21:12:51 one minetestserver[362992]: 2022-06-28 21:12:51: ACTION[Server]: [doc] Wrote player data into /var/lib/minetest/minetest//doc.mt.
Jun 28 21:12:51 one minetestserver[362992]: 2022-06-28 21:12:51: ACTION[Server]: cy leaves game. List of players: straw
Jun 28 21:12:51 one minetestserver[362992]: 2022-06-28 21:12:51: WARNING[Server]: ScriptApiBase::objectrefGetOrCreate(): Pushing ObjectRef to removed/deactivated object, this is probably a bug.
Jun 28 21:12:54 one systemd[1]: minetest@minetest.service: Main process exited, code=dumped, status=11/SEGV
Jun 28 21:12:54 one systemd[1]: minetest@minetest.service: Failed with result 'core-dump'.
Jun 28 21:12:54 one systemd[1]: minetest@minetest.service: Consumed 57min 11.327s CPU time.
```
It's running MineClone5 with the ts_furniture mod (which is missing the `default` dependency for some reason and therefore isn't enabled, pretty sure this happened prior to adding it though)

Screenshot of the CPU spike immediately prior to the crash on `btm`:
![image](https://user-images.githubusercontent.com/45886986/176294364-0f03c8a6-64fa-41c4-a695-df81ab10aa9e.png)

##### Steps to reproduce

My Minetest server runs MineClone5 only, along with ts_furniture which I haven't enabled

**Comment by sfan5** [comment]
Without a backtrace this is unfortunately not very meaningful, you can try `coredumpctl dump` to see if/what information systemd has automatically recorded for this crash.

**Comment by acoolstraw** [comment]
Output of `coredumpctl dump` wouldn't be sensitive, right?

**Comment by acoolstraw** [comment]
here, anyway:
```
           PID: 362992 (minetestserver)
           UID: 967 (minetest)
           GID: 967 (minetest)
        Signal: 11 (SEGV)
     Timestamp: Tue 2022-06-28 21:12:51 UTC (3h 42min ago)
  Command Line: /usr/bin/minetestserver --config /etc/minetest/minetest.conf --map-dir /var/lib/minetest/minetest/
    Executable: /usr/bin/minetestserver
 Control Group: /system.slice/system-minetest.slice/minetest@minetest.service
          Unit: minetest@minetest.service
         Slice: system-minetest.slice
       Boot ID: [redacted?]
    Machine ID: [redacted?]
      Hostname: one
       Storage: /var/lib/systemd/coredump/core.minetestserver.967.233353a0ed9e438b81795ae0bea24ed5.362992.1656450771000000.zst (present)
     Disk Size: 82.7M
       Message: Process 362992 (minetestserver) of user 967 dumped core.
                
                Module linux-vdso.so.1 with build-id 20e297295de7b5f8538cab53afe224f81700405b
                Module libnss_resolve.so.2 with build-id 1790ccea3fad28777fc366683fe31ed6581c9446
                Module libcap.so.2 with build-id 21b00ab76bb16fde56fd741f07233c78069f0579
                Module libnss_mymachines.so.2 with build-id 51b4a7127dac7bcef5aa6e3046e0509069dcba00
                Module libbrotlicommon.so.1 with build-id acfd597a977c8087bb6184383daae2e828a9ce42
                Module libunistring.so.2 with build-id 617dbf3d3d6f85d6556a7a036e23845e95490158
                Module libsasl2.so.3 with build-id 0c10b132405d1c16805173607ae5496c9bc84ba7
                Module liblber.so.2 with build-id c3d8f0b19222aef1692d78535f3e795760bf07a1
                Module libresolv.so.2 with build-id 89a368a6ad1b392d126a2a5beb9c2f61ade00279
                Module libkeyutils.so.1 with build-id ac405ddd17be10ce538da3211415ee50c8f8df79
                Module libkrb5support.so.0 with build-id 15f223925ef59dee4379ebbc0fcd14eda9ba81a2
                Module libcom_err.so.2 with build-id 3360a28740ffbbd5a5c0c21d09072445908707e5
                Module libk5crypto.so.3 with build-id cc77a742cb62447a53d98285b41558b8acd92866
                Module libkrb5.so.3 with build-id 371cc767dacb17cb42c9c44b88eebbed5ee9a756
                Module libbrotlidec.so.1 with build-id 66c54e9301f7e102ecc1d88547e5f0e8a056fe22
                Module libpsl.so.5 with build-id 0229a201aaf5652186c9fdc192ebe52baf19d7f1
                Module libssh2.so.1 with build-id a4adfe44cc7ebd295b3b783361acc3dcfcea1d50
                Module libidn2.so.0 with build-id 1ce2b50ad9f9821c2c629b521cf5a3c99593d332
                Module libnghttp2.so.14 with build-id 16f0981d5251b03b11a49236ac403562ee458887
                Module libpthread.so.0 with build-id 95ae4f30a6f12ccbff645d30f8e1a3ee23ec7d36
                Module libsnappy.so.1 with build-id 36e3fb247a476fe2f755162644ebcd8ebd5d92cb
                Module libldap.so.2 with build-id c9be4f0aa087ad90d69a828c34413da4f1ee3ea3
                Module libgssapi_krb5.so.2 with build-id 292f1ce32161c0ecc4a287bc8494d5d7c420a03f
                Module libcrypto.so.1.1 with build-id d1f36af479cd3316f5ea2460b330fbe703587f12
                Module libssl.so.1.1 with build-id e6b1f97a5b60b4248c49dfc5b11f53f281b507d0
                Module ld-linux-x86-64.so.2 with build-id 0effd0e43efa4468d3c31871c93af0b7f3005673
                Module libc.so.6 with build-id 60df1df31f02a7b23da83e8ef923359885b81492
                Module libgcc_s.so.1 with build-id 0e3de903950e35ae59a5de8c00b1817a4a71ca01
                Module libm.so.6 with build-id 1b7296ef9fd806e47060788389293c824b09ad72
                Module libstdc++.so.6 with build-id b3eaebcb5369487562e52dd8d016c9cd77f80c54
                Module libcurl.so.4 with build-id a332bbdef106a70e40f82ae7015932a5c9c9befe
                Module libspatialindex.so.6 with build-id a75ad940d6e1df68fa57520fe9bf8316379d24e8
                Module libhiredis.so.1.0.0 with build-id 2421bd3130ca487cdf1a55d1569bb69e73b74c79
                Module libleveldb.so.1 with build-id a5d4a728885383f1bee4e4c242cb2975dbec63b8
                Module libpq.so.5 with build-id f7450841da29c58f579f145ba228c39a372d1e7b
                Module libncursesw.so.6 with build-id b9917757481e6fa6097e2a1f31f5bb5eaf138c4e
                Module libgmp.so.10 with build-id 26cec2ebe94cc5c4cb99e6988717347222b324fd
                Module libluajit-5.1.so.2 with build-id b0dc62978053c6564fee1c6f270f37a60c6f683e
                Module libjsoncpp.so.25 with build-id 4b8bb6b1d3b6881982a8e7e3f1bffd8130f6ae54
                Module libsqlite3.so.0 with build-id 143cb4bde0356249e42c865273631c175d738ba1
                Module libzstd.so.1 with build-id ab54c2881f53ab314e134f3e08c76d504376dd5d
                Module libz.so.1 with build-id fefe3219a96d682ec98fcfb78866b8594298b5a2
                Module minetestserver with build-id e4c9d529bed7d52a4101eb7f591b34480d5b57af
                Stack trace of thread 363002:
                #0  0x0000557442d20647 n/a (minetestserver + 0x1c6647)
                #1  0x0000557442cda413 n/a (minetestserver + 0x180413)
                #2  0x0000557442c9af6a n/a (minetestserver + 0x140f6a)
                #3  0x00007f1f795f9f6b n/a (libluajit-5.1.so.2 + 0x9f6b)
                #4  0x00007f1f7960c843 lua_pcall (libluajit-5.1.so.2 + 0x1c843)
                #5  0x0000557442d0fec2 n/a (minetestserver + 0x1b5ec2)
                #6  0x0000557442e319ad n/a (minetestserver + 0x2d79ad)
                #7  0x0000557442d0d8a6 n/a (minetestserver + 0x1b38a6)
                #8  0x0000557442e2c441 n/a (minetestserver + 0x2d2441)
                #9  0x0000557442e158ee n/a (minetestserver + 0x2bb8ee)
                #10 0x0000557442e07aa2 n/a (minetestserver + 0x2adaa2)
                #11 0x0000557442d21c97 n/a (minetestserver + 0x1c7c97)
                #12 0x00007f1f79118183 execute_native_thread_routine (libstdc++.so.6 + 0xd6183)
                #13 0x00007f1f78dbb54d n/a (libc.so.6 + 0x8c54d)
                #14 0x00007f1f78e40874 __clone (libc.so.6 + 0x111874)
                
                Stack trace of thread 362992:
                #0  0x00007f1f78e06fd5 clock_nanosleep (libc.so.6 + 0xd7fd5)
                #1  0x00007f1f78e0bb97 __nanosleep (libc.so.6 + 0xdcb97)
                #2  0x00007f1f78e37d59 usleep (libc.so.6 + 0x108d59)
                #3  0x0000557442e200b7 n/a (minetestserver + 0x2c60b7)
                #4  0x0000557442dd0004 n/a (minetestserver + 0x276004)
                #5  0x0000557442bdb327 n/a (minetestserver + 0x81327)
                #6  0x00007f1f78d58290 n/a (libc.so.6 + 0x29290)
                #7  0x00007f1f78d5834a __libc_start_main (libc.so.6 + 0x2934a)
                #8  0x0000557442beb035 n/a (minetestserver + 0x91035)
                
                Stack trace of thread 363003:
                #0  0x00007f1f78db8119 n/a (libc.so.6 + 0x89119)
                #1  0x00007f1f78dc316b n/a (libc.so.6 + 0x9416b)
                #2  0x0000557442da7e02 n/a (minetestserver + 0x24de02)
                #3  0x0000557442d21c97 n/a (minetestserver + 0x1c7c97)
                #4  0x00007f1f79118183 execute_native_thread_routine (libstdc++.so.6 + 0xd6183)
                #5  0x00007f1f78dbb54d n/a (libc.so.6 + 0x8c54d)
                #6  0x00007f1f78e40874 __clone (libc.so.6 + 0x111874)
                
                Stack trace of thread 363004:
                #0  0x00007f1f78db8119 n/a (libc.so.6 + 0x89119)
                #1  0x00007f1f78dba920 pthread_cond_wait (libc.so.6 + 0x8b920)
                #2  0x00007f1f79110d71 __gthread_cond_wait (libstdc++.so.6 + 0xced71)
                #3  0x0000557442d9f16c n/a (minetestserver + 0x24516c)
                #4  0x0000557442d21c97 n/a (minetestserver + 0x1c7c97)
                #5  0x00007f1f79118183 execute_native_thread_routine (libstdc++.so.6 + 0xd6183)
                #6  0x00007f1f78dbb54d n/a (libc.so.6 + 0x8c54d)
                #7  0x00007f1f78e40874 __clone (libc.so.6 + 0x111874)
                
                Stack trace of thread 362994:
                #0  0x00007f1f78e3747c __select (libc.so.6 + 0x10847c)
                #1  0x0000557442c93276 n/a (minetestserver + 0x139276)
                #2  0x0000557442c783dd n/a (minetestserver + 0x11e3dd)
                #3  0x0000557442c7a523 n/a (minetestserver + 0x120523)
                #4  0x0000557442d21c97 n/a (minetestserver + 0x1c7c97)
                #5  0x00007f1f79118183 execute_native_thread_routine (libstdc++.so.6 + 0xd6183)
                #6  0x00007f1f78dbb54d n/a (libc.so.6 + 0x8c54d)
                #7  0x00007f1f78e40874 __clone (libc.so.6 + 0x111874)
                
                Stack trace of thread 362993:
                #0  0x00007f1f78db8119 n/a (libc.so.6 + 0x89119)
                #1  0x00007f1f78dc316b n/a (libc.so.6 + 0x9416b)
                #2  0x0000557442c6a4fc n/a (minetestserver + 0x1104fc)
                #3  0x0000557442d21c97 n/a (minetestserver + 0x1c7c97)
                #4  0x00007f1f79118183 execute_native_thread_routine (libstdc++.so.6 + 0xd6183)
                #5  0x00007f1f78dbb54d n/a (libc.so.6 + 0x8c54d)
                #6  0x00007f1f78e40874 __clone (libc.so.6 + 0x111874)
                ELF object binary architecture: AMD x86-64
Refusing to dump core to tty (use shell redirection or specify --output).
```

**Comment by debagos** [comment]
> SEGV
> LuaJIT 2.1.0-beta3

I am pretty sure LuaJIT is the problem here. I had a lot segfaults using LuaJIT.
Switch to the builtin Lua and see if the server is stable again.

Rant:
Some LuaJIT versions gave me multiple random crashes per day.
Others crashed only when someone did a specific thing.
And some other versions again only crashed very occasionally, like once in a week.
Even the most recent version already crashed once on my server, which isn't too bad for LuaJIT standards, because the server is running 24/7 with a few players during the day and the current version is running for 10 days now.

LuaJIT is blazing fast, compared with the bundled Lua, but it's a big effin unstable mess.

PS:
Don't forget to tell us how it went, thanks.

**Comment by sfan5** [comment]
Windows builds have been shipping with LuaJIT 2.1.0-beta3 for years with no known issues traced or suspected to be LuaJIT bugs.
Confusingly enough Arch Linux (what OP is using) ships LuaJIT from git, despite identical version number, so you should make sure you have the latest version.

In the end it's not sure to say whether a crash is caused by LuaJIT without analyzing a backtrace, we don't have one here (yet).

**Comment by appgurueu** [comment]
I ran into a LuaJIT bug while fuzzing my serializers. Switching to LuaJIT latest master fixed it.

**Comment by debagos** [comment]
The true `2.1.0-beta3` version got released in May 2017 and is very stable, never crashed me.
But the maintainer changed a lot in that version since then, see https://github.com/LuaJIT/LuaJIT
> Windows builds have been shipping with LuaJIT 2.1.0-beta3 for years

And obviously you compile MT against the original release (which I strongly suggest you keep doing).
So your experience with LuaJIT is a different one than mine, because we use different versions.
> Arch Linux ships LuaJIT from git

And I told you my experience with the recent versions...
Still remember this one: #12163 ?
Before I opened that issue I had SIGIOTs for months, never was able to find the reason. After you guys helped me find the culprit LuaJIT, I checked all logs and most of my crashes were because of LuaJIT (over multiple versions), not mods. Pretty telling.
Oh and check this out to get a feeling how often LuaJIT is broken, you will see it's not just warnings and light stuff that needed to get fixed:
https://github.com/search?q=repo%3ALuaJIT%2FLuaJIT+fix&type=Commits
> CPU spikes prior to the crash

I noticed that too at times by the way.
The way he describes the crashes reminds me of typical LuaJIT crashes, that's why I suggested to try switching to the builtin version.
But yes, we need a backtrace, no doubt.
Anyway, I would bet my two cents that LuaJIT is the culprit here.
> `Pushing ObjectRef to removed/deactivated object, this is probably a bug.`
>is worth investigating. MineClone + Minetest is more likely to be the issue here rather than LuaJIT

I get this messages often during shutdowns of my server. I think this "bug" happened after the shutdown started.
